### PR TITLE
Add LogicSig templates and example code.

### PR DESCRIPTION
**Pull request by derbear**
Let's fast-track merging this so that end-to-end tests can start getting written for our features.

The following things are TODO:

1. Changing `driver.py` to support command-line options more consistently with existing `goal` tools. 
2. Integrating the custom signature tool `dsign` with the other commands.
3. Improving the documentation for the templates so that it's more obvious how to use them.
4. Improving efficiency of the template code (reverse argument push order for long conjunctions; fix `pop` branch bug)
5. Implement an oracle-controlled escrow.

### CODE REVIEW

File: cmd/goal/clerk.go [hunk 1]
```diff
@@ -867,7 +868,7 @@ var dryrunCmd = &cobra.Command{
 		block := bookkeeping.Block{}
 		block.BlockHeader.Round = basics.Round(round)
 		block.BlockHeader.TimeStamp = timeStamp
-		proto := config.Consensus[protocol.ConsensusCurrentVersion]
+		proto := config.Consensus[protocol.ConsensusFuture]
```
Comment (pzbitskiy) [inline cmd/goal/clerk.go]: do we want this in master

Comment (derbear) [reply cmd/goal/clerk.go]: Maybe we want to wait until the Foundation gives us the right spec name?

File: data/transactions/logic/assembler.go [hunk 2]
```diff
@@ -680,10 +685,14 @@ func (ops *OpStream) Assemble(fin io.Reader) error {
...
 			// create a label
-			ops.SetLabelHere(opstring[:len(opstring)-1])
+			err := ops.SetLabelHere(opstring[:len(opstring)-1])
+			if err != nil {
+				return lineErr(ops.sourceLine, err)
+			}
 			continue
 		}
-		return fmt.Errorf(":%d unknown opcode %v", ops.sourceLine, opstring)
+		err := fmt.Errorf("unknown opcode %v", opstring)
```
[the remark this answers is not in the document]
Comment (algobolson) [reply data/transactions/logic/assembler.go]: going the other way from $s, %#v will print a string with escaping so you can know exactly what it was

**Comment by stechu** [comment]
LGTM.

File: data/transactions/logic/assembler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (derbear) [reply data/transactions/logic/assembler.go]: Will defer to a future PR.

**Pull request merged**